### PR TITLE
feat: bring up Array.shuffle

### LINKED ISSUE #100: Implement Array.shuffle
To return same array but with randomized order of elements.

Some guide https://stackoverflow.com/questions/2450954/how-to-randomize-shuffle-a-javascript-array

### PR DESCRIPTION
**Pull request by SukkaW**
The PR closes #100, bringing up a side-effect-free, and O(n) array shuffle function.

The unit test is not finished since `tad` is not powerful enough to write a unit test for `shuffle` (Help is wanted).

P.S. I want to adopt the unit test from https://github.com/philihp/fast-shuffle/blob/master/src/__tests__/index.test.js

### CODE REVIEW

File: array/shuffle.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+	var sourceIndex = input.length;
+	var destinationIndex = 0;
+
+	// Pre-assign the size of array
+	var shuffled = new Array(sourceIndex);
+
+	while (sourceIndex) {
+		// eslint-disable-next-line no-bitwise
+		var randomIndex = (sourceIndex * random()) | 0;
```
[the remark this answers is not in the document]
Comment (medikoo) [reply array/shuffle.js:20]: Thanks, still this benchmark doesn't seem to test purely `Math.round` vs bitwise stripping.

I've prepared one here: https://jsbench.me/mske89xfr3/1 and it shows negligible difference between both (note that both operations on its own are extremely fast)

It seems not harmful to stick to `Math.round` here, and it'll definitely be win for readability

File: array/shuffle.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	if (!isArray(input)) throw new TypeError("input must be an Array, got " + typeof input + "!");
+
+	// Make a clone of original array
+	var $array = input.slice(0);
+
+	var sourceIndex = input.length;
+	var destinationIndex = 0;
+
+	// Pre-assign the size of array
+	var shuffled = new Array(sourceIndex);
```
[the remark this answers is not in the document]
Comment (medikoo) [reply array/shuffle.js:16]: Thanks for explanation!

**Review by medikoo — COMMENTED** [review]
Thank you @SukkaW it looks great! I have just one (more a readability) concern with using bitwise notation. see my comment.

File: array/shuffle.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (SukkaW) [reply array/shuffle.js:20]: I have updated the original benchmark with `Math.floor`: https://repl.it/repls/CornyBrightConditions

It seems that `Math.floor` is still slower (At least it is slower in Node.js)

[the remark this answers is not in the document]
Comment (medikoo) [reply array/shuffle.js:20]: This benchmark still involves a lot more code than it should. Technically it's purely about testing `Math.round(value)` against `value | 0`.

Node.js uses same engine as Chrome, and difference that https://jsbench.me/mske89xfr3/1 shows over there is neglible.

Also this package is environment agnostic, meant to be used in any engine (e.g. in Firefox there's no speed difference between both)

[the remark this answers is not in the document]
Comment (medikoo) [reply array/shuffle.js:20]: I'd use `Math.floor` now, and eventually revisit if we would have a real world report that it noticeable slows down performance somewhere.

**Pull request closed without merging**